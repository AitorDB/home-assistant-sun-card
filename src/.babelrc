{
  "presets": [
    "@babel/env",
    "@babel/typescript"
  ],
  "plugins": [
    ["@babel/plugin-proposal-decorators", { "decoratorsBeforeExport": true }],
    "@babel/proposal-class-properties"
  ]
}
