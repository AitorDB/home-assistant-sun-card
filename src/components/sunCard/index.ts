export * from './SunCard'
export * from './SunCardContent'
export * from './SunCardFooter'
export * from './SunCardGraph'
export * from './SunCardHeader'
