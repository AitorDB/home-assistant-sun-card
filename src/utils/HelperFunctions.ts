import { html, TemplateResult } from 'lit-html'
import { Constants } from '../constants'
import { I18N } from './I18N'

export class HelperFunctions {
  public static nothing (): TemplateResult {
    return html``
  }

  public static renderFieldElement (i18n: I18N, translationKey: string, value: Date | number | string | undefined): TemplateResult {
    let display: string
    if (value === undefined) {
      return HelperFunctions.nothing()
    } else if (value instanceof Date) {
      display = i18n.formatDateAsTime(value)
    } else {
      display = value.toString()
    }

    return html`
      <div class="sun-card-text-container">
        <span class="sun-card-field-name">${ i18n.tr(translationKey) }</span>
        <span class="sun-card-field-value">${ display }</span>
      </div>
    `
  }
  
  public static isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  }

  public static todayAtStartOfDay (): Date {
    const today = new Date()
    today.setHours(0)
    today.setMinutes(0)
    today.setSeconds(0)
    today.setMilliseconds(0)
  
    return today
  }
  
  public static todayAtEndOfDay (): Date {
    const today = new Date()
    today.setHours(23)
    today.setMinutes(59)
    today.setSeconds(59)
    today.setMilliseconds(999)
  
    return today
  }

  public static findSectionPosition (msSinceSectionStart: number, msSectionEnd: number, section: number): number {    
    return (Math.min(msSinceSectionStart, msSectionEnd) * section) / msSectionEnd
  }

  public static findSunProgress (sunPosition: number, startPosition: number, endPosition: number): number {
    return HelperFunctions.clamp(0, 100,
      (100 * (sunPosition - startPosition)) / (endPosition - startPosition)
    )
  }
  public static clamp (min: number, max: number, value: number): number {
    if (min === max) {
      return min
    }

    if (min > max) {
      throw new RangeError('Min value can not be bigger than the max value')
    }

    return Math.min(Math.max(value, min), max)
  }
}
