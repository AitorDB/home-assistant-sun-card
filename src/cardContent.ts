import { html, TemplateResult } from 'lit-element'
import { TSunCardConfig, TSunCardData, TSunCardTexts, TSunCardTime } from './types'

export class SunCardContent {
  static generate (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    return html`
      <ha-card>
        <div class="sun-card ${config.darkMode ? '' : 'sun-card-light'}">
          ${this.generateHeader(data, localization, config)}
          ${this.generateBody(data)}
          ${this.generateFooter(data, localization, config)}
        </div>
      </ha-card>
    `
  }

  private static generateHeader (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    const title = config.title !== undefined ? html`
      <h1 class="sun-card-title">${config.title}</h1>
    ` : html``

    return html`
      ${title}
      <div class="sun-card-header">
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Sunrise}</span>
          ${data?.times.sunrise ? this.generateTime(data.times.sunrise) : ''}

        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Sunset}</span>
          ${data?.times.sunset ? this.generateTime(data.times.dawn) : ''}
        </div>
      </div>
    `
  }

  private static generateBody (data: TSunCardData): TemplateResult {
    const sunID = Math.random().toString(36).replace('0.', '')
    const dawnID = Math.random().toString(36).replace('0.', '')
    const dayID = Math.random().toString(36).replace('0.', '')
    const duskID = Math.random().toString(36).replace('0.', '')

    return html`
      <div class="sun-card-body">
        <svg viewBox="0 0 550 150" xmlns="http://www.w3.org/2000/svg">
          <defs>
            <linearGradient id="${sunID}" x1="0%" y1="0%" x2="0%" y2="100%">
              <stop offset="0%" style="stop-color:#f9d05e;stop-opacity:1" />
              <stop offset="${data?.sunPercentOverHorizon ?? 0}%" style="stop-color:#f9d05e;stop-opacity:1" />
              <stop offset="${data?.sunPercentOverHorizon ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${dawnID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.dawnProgressPercent ?? 0}%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.dawnProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${dayID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#8ebeeb;stop-opacity:1" />
              <stop offset="${data?.dayProgressPercent ?? 0}%" style="stop-color:#8ebeeb;stop-opacity:1" />
              <stop offset="${data?.dayProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
            
            <linearGradient id="${duskID}" x1="0%" y1="0%" x2="100%" y2="0%">
              <stop offset="0%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.duskProgressPercent ?? 0}%" style="stop-color:#393b78;stop-opacity:1" />
              <stop offset="${data?.duskProgressPercent ?? 0}%" style="stop-color:rgb(0,0,0,0);stop-opacity:1" />
            </linearGradient>
          </defs>
          <path class="sun-card-sun-line" d="M5,146 C29,153 73,128 101,108 C276,-29 342,23 449,108 C473,123 509,150 545,146" fill="none" stroke="var(--sun-card-lines)" shape-rendering="geometricPrecision" />
          <path d="M5,146 C29,153 73,128 101,108 L 5 108" fill="url(#${dawnID})" opacity="${data?.dawnProgressPercent ? 1 : 0}" stroke="url(#${dawnID})" shape-rendering="geometricPrecision" />
          <path d="M101,108 C276,-29 342,23 449,108 L 104,108" fill="url(#${dayID})" opacity="${data?.dayProgressPercent ? 1 : 0}" stroke="url(#${dayID})" shape-rendering="geometricPrecision" />
          <path d="M449,108 C473,123 509,150 545,146 L 545 108" fill="url(#${duskID})" opacity="${data?.duskProgressPercent ? 1 : 0}" stroke="url(#${duskID})" shape-rendering="geometricPrecision" />
          <line x1="5" y1="108" x2="545" y2="108" stroke="var(--sun-card-lines)" />
          <line x1="101" y1="25" x2="101" y2="100" stroke="var(--sun-card-lines)" />
          <line x1="449" y1="25" x2="449" y2="100" stroke="var(--sun-card-lines)" />
          <circle cx="${data?.sunPosition.x ?? 0}" cy="${data?.sunPosition.y ?? 0}" r="17" opacity="${data?.sunPercentOverHorizon ? 1 : 0}" stroke="none" fill="url(#${sunID})" shape-rendering="geometricPrecision" />
        </svg>
      </div>
    `
  }

  private static generateFooter (data: TSunCardData, localization: TSunCardTexts, config: TSunCardConfig): TemplateResult {
    const upperRow = html`
      <div class="sun-card-footer-row">
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Dawn}</span>
          ${data?.times.dawn ? this.generateTime(data.times.dawn) : ''}
        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Noon}</span>
          ${data?.times.noon ? this.generateTime(data.times.noon) : ''}
        </div>
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Dusk}</span>
          ${data?.times.dusk ? this.generateTime(data.times.dusk) : ''}
        </div>
      </div>
    `

    let bottomRow = html``
    if (config.showAzimuth || config.showElevation) {
      const azimuth = config.showAzimuth ? html`
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Azimuth}</span>
          <span class="sun-card-dawn-time sun-card-text-time">${data?.azimuth ?? ''}</span>
        </div>
      ` : html``

      const elevation = config.showElevation ? html`
        <div class="sun-card-text-container">
          <span class="sun-card-text-subtitle">${localization.Elevation}</span>
          <span class="sun-card-dawn-time sun-card-text-time">${data?.elevation ?? ''}</span>
        </div>
      ` : html``
  
      bottomRow = html`
        <div class="sun-card-footer-row">
          ${azimuth}
          ${elevation}
        </div>
      `
    }

    return html`
      <div class="sun-card-footer">
        ${upperRow}
        ${bottomRow}
      </div>
    `
  }

  private static generateTime (time: TSunCardTime) {
    if (time.period) {
      return html`
        <span class="sun-card-text-time">
          ${time.time} <span class="sun-card-text-time-period">${time.period}</span>
        </span>
      `
    }
    
    return html`
      <span class="sun-card-text-time">${time.time}</span>
    `
  }
}
