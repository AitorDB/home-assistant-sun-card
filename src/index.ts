import { HomeAssistant } from "custom-card-helpers";
import { customElement, LitElement, state } from "lit-element";

import cardStyles from "./cardStyles";
import { Constants } from "./constants";
import { SunCardContent } from "./cardContent";
import { TSunCardConfig, TSunCardData, TSunCardTime } from "./types";

@customElement("sun-card")
class SunCard extends LitElement {
  static readonly cardType = "sun-card";
  static readonly cardName = "Sun Card";
  static readonly cardDescription =
    "Custom card that display a graph to track the sun position and related events";

  @state()
  private config: TSunCardConfig = {};

  @state()
  private data!: TSunCardData;

  private hasRendered = false;
  private lastHass!: HomeAssistant;

  set hass(hass: HomeAssistant) {
    this.lastHass = hass;

    if (!this.hasRendered) {
      return;
    }

    this.processLastHass();
  }

  calculatePositionAndProgressesByTime(hass: HomeAssistant) {
    const sunLine = this.shadowRoot?.querySelector("path") as SVGPathElement;
    const sunrise = new Date(hass.states["sun.sun"].attributes.next_rising);
    const sunset = new Date(hass.states["sun.sun"].attributes.next_setting);
    const eventsAt = {
      dayStart: 0,
      sunrise: this.convertDateToMinutesSinceDayStarted(sunrise),
      sunset: this.convertDateToMinutesSinceDayStarted(sunset),
      dayEnd: 23 * 60 + 59,
    };

    const now = new Date();
    const minutesSinceTodayStarted =
      this.convertDateToMinutesSinceDayStarted(now);

    // Dawn section position [0 - 105]
    const dawnSectionPosition =
      (Math.min(minutesSinceTodayStarted, eventsAt.sunrise) * 105) /
      eventsAt.sunrise;

    // Day section position [106 - 499]
    const minutesSinceDayStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunrise,
      0
    );
    const daySectionPosition =
      (Math.min(minutesSinceDayStarted, eventsAt.sunset - eventsAt.sunrise) *
        (499 - 106)) /
      (eventsAt.sunset - eventsAt.sunrise);

    // Dusk section position [500 - 605]
    const minutesSinceDuskStarted = Math.max(
      minutesSinceTodayStarted - eventsAt.sunset,
      0
    );
    const duskSectionPosition =
      (minutesSinceDuskStarted * (605 - 500)) /
      (eventsAt.dayEnd - eventsAt.sunset);

    const position =
      dawnSectionPosition + daySectionPosition + duskSectionPosition;
    const sunPosition = sunLine.getPointAtLength(position);

    const dawnProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.dayStart)) /
      (Constants.EVENT_X_POSITIONS.sunrise -
        Constants.EVENT_X_POSITIONS.dayStart);
    const dayProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunrise)) /
      (Constants.EVENT_X_POSITIONS.sunset -
        Constants.EVENT_X_POSITIONS.sunrise);
    const duskProgressPercent =
      (100 * (sunPosition.x - Constants.EVENT_X_POSITIONS.sunset)) /
      (Constants.EVENT_X_POSITIONS.dayEnd - Constants.EVENT_X_POSITIONS.sunset);

    const sunYTop = sunPosition.y - Constants.SUN_RADIUS;
    const yOver = Constants.HORIZON_Y - sunYTop;
    let sunPercentOverHorizon = 0;
    if (yOver > 0) {
      sunPercentOverHorizon = Math.min(
        (100 * yOver) / (2 * Constants.SUN_RADIUS),
        100
      );
    }

    return {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition: { x: sunPosition.x, y: sunPosition.y },
    };
  }

  convertDateToMinutesSinceDayStarted(date: Date) {
    return date.getHours() * 60 + date.getMinutes();
  }

  parseTime(timeText: string, locale?: string) {
    const regex = /\d{1,2}[:.]\d{1,2}|[AMP]+/g;
    const date = new Date(timeText);
    const { language, timeFormat } = this.getConfig();
    const result = date
      .toLocaleTimeString(locale ?? language, { hour12: timeFormat === "12h" })
      .match(regex) as [string, ("AM" | "PM")?];

    if (!result && !locale) {
      return this.parseTime(timeText, Constants.DEFAULT_CONFIG.language);
    }

    const [time, period] = result;
    return { time, period };
  }
  calculateTimeBetweenDates(date1: Date, date2: Date): TSunCardTime {
    const timeMs = Math.abs(date1 - date2);
    const { language } = this.getConfig();
    const time = new Date(timeMs).toLocaleTimeString(language, {
      hour12: false,
    });
    return { time };
  }
  processLastHass() {
    if (!this.lastHass) {
      return;
    }

    this.config.darkMode =
      this.config.darkMode ?? this.lastHass.themes.darkMode;
    this.config.language =
      this.config.language ??
      this.lastHass.locale?.language ??
      this.lastHass.language;
    this.config.timeFormat =
      this.config.timeFormat ??
      this.getTimeFormatByLanguage(this.config.language);

    const times = {
      dawn: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dawn
      ),
      dusk: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_dusk
      ),
      noon: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_noon
      ),
      sunrise: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_rising
      ),
      sunset: this.parseTime(
        this.lastHass.states["sun.sun"].attributes.next_setting
      ),
    };

    const {
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      sunPercentOverHorizon,
      sunPosition,
    } = this.calculatePositionAndProgressesByTime(this.lastHass);

    const lengthOfDay = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_rising),
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting)
    );

    const daylightLeft = this.calculateTimeBetweenDates(
      new Date(this.lastHass.states["sun.sun"].attributes.next_setting),
      new Date()
    );

    const data: TSunCardData = {
      azimuth: this.lastHass.states["sun.sun"].attributes.azimuth,
      dawnProgressPercent,
      dayProgressPercent,
      duskProgressPercent,
      elevation: this.lastHass.states["sun.sun"].attributes.elevation,
      sunPercentOverHorizon,
      sunPosition,
      times,
      lengthOfDay,
      daylightLeft,
    };

    this.data = data;
  }

  getConfig() {
    const config: TSunCardConfig = {};
    config.darkMode = this.config.darkMode ?? Constants.DEFAULT_CONFIG.darkMode;
    config.language = this.config.language ?? Constants.DEFAULT_CONFIG.language;
    config.showAzimuth =
      this.config.showAzimuth ?? Constants.DEFAULT_CONFIG.showAzimuth;
    config.showElevation =
      this.config.showElevation ?? Constants.DEFAULT_CONFIG.showElevation;
    config.timeFormat =
      this.config.timeFormat ?? Constants.DEFAULT_CONFIG.timeFormat;
    config.title = this.config.title;

    if (
      !Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(config.language!)
    ) {
      config.language = Constants.DEFAULT_CONFIG.language;
    }

    return config;
  }

  getTimeFormatByLanguage(language: string) {
    const date = new Date();
    const time = date.toLocaleTimeString(language).toLocaleLowerCase();
    return time.includes("pm") || time.includes("am") ? "12h" : "24h";
  }

  setConfig(config: TSunCardConfig) {
    this.config = { ...config };
  }

  protected render() {
    const config = this.getConfig();
    const language = config.language!;
    const localization = Constants.LOCALIZATION_LANGUAGES[language];
    return SunCardContent.generate(this.data, localization, config);
  }

  protected updated(changedProperties) {
    super.updated(changedProperties);

    if (!this.hasRendered) {
      this.hasRendered = true;
      this.processLastHass();
    }
  }

  static get styles() {
    return cardStyles;
  }
}

window.customCards = window.customCards || [];
window.customCards.push({
  type: SunCard.cardType,
  name: SunCard.cardName,
  description: SunCard.cardDescription,
});
