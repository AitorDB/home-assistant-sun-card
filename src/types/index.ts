import { LovelaceCardConfig } from 'custom-card-helpers'
import { I18N } from '../utils/I18N'


export type TSunCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface ISunCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string
  
  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: TSunCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number
  
  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum ESunCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type TSunCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: TSunCardTimes

  error?: ESunCardErrors
}

export type TSunCardTimes = {
  dawn: Date
  dusk: Date
  noon: Date
  sunrise: Date
  sunset: Date
}

export enum ESunCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type TSunCardI18NKeys = { [key in ESunCardI18NKeys ]: string } | { errors: TSunCardI18NErrorKeys }

export type TSunCardI18N = Record<string, unknown>

export type TSunCardI18NErrorKeys = {
  [key in ESunCardErrors]: string
}
