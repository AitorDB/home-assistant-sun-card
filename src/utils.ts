import { html, TemplateResult } from 'lit-html'
import { Constants } from './constants'
import { TSunCardI18N, TSunCardI18NKeys } from './types'

export class I18N {
  private localization: TSunCardI18NKeys
  private dateFormatter: Intl.DateTimeFormat

  constructor (language: string, use12HourClock: boolean) {
    this.localization = Constants.LOCALIZATION_LANGUAGES[language]

    // https://developer.mozill a.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat/DateTimeFormat
    this.dateFormatter = new Intl.DateTimeFormat(language, {
      hour12: use12HourClock,
      timeStyle: 'short'
    })
  }

  public formatDateAsTime (date: Date): string {
    return this.dateFormatter.formatToParts(date).map(({ type, value }) => {
      switch (type) {
        // intentional fallthrough
        case 'hour':
        case 'minute':
        case 'dayPeriod':
        case 'literal':
          return value
        default:
          return ''
      }
    })
    .join('')
  }

  public tr (trKey: string): string {
    return this.getLocalizationElement(this.localization, trKey).toString()
  }

  // Janky recursive logic to handle nested values in i18n json sources
  private getLocalizationElement (localization: TSunCardI18N, trKey: string): string | TSunCardI18N {

    if (trKey.includes('.')) {
      const parts = trKey.split('.', 2)
      // TODO: maybe add typecheck
      const localization = this.getLocalizationElement(this.localization, parts[0]) as TSunCardI18N
      return this.getLocalizationElement(localization, parts[1])
    } else {
      // if the translation isn't completed in the target language, fall back to english
      // give ugly string for developers who misstype
      return (localization ? localization[trKey] : undefined)
        ?? Constants.FALLBACK_LOCALIZATION[trKey]
        ?? `Translation key '${trKey}' doesn't have a valid translation`
    }
  }
}

export default {
  nothing: html``,
  renderFieldElement (i18n: I18N, translationKey: string, value: Date | number | string): TemplateResult {
    
    let display: string
    if (value instanceof Date) {
      display = i18n.formatDateAsTime(value)
    } else {
      display = value.toString()
    }

    return html`
      <div class="sun-card-text-container">
        <span class="sun-card-field-name">${ i18n.tr(translationKey) }</span>
        <span class="sun-card-field-value">${ display }</span>
      </div>
    `
  },
  isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  },
  todayAtStartOfDay (): Date {
    const today = new Date()
    today.setHours(0)
    today.setMinutes(0)
    today.setSeconds(0)
    today.setMilliseconds(0)
  
    return today
  },
  todayAtEndOfDay (): Date {
    const today = new Date()
    today.setHours(23)
    today.setMinutes(59)
    today.setSeconds(59)
    today.setMilliseconds(999)
  
    return today
  },
  findSectionPosition (msSinceSectionStart: number, msSectionEnd: number, section: number): number {    
    return (Math.min(msSinceSectionStart, msSectionEnd) * section) / msSectionEnd
  },
  findSunProgress (sunPosition: number, startPosition: number, endPosition: number): number {
    return this.clamp(0, 100,
      (100 * (sunPosition - startPosition)) / (endPosition - startPosition)
    )
  },
  clamp (min: number, max: number, value: number): number {
    return Math.min(Math.max(value, min), max)
  }
}
