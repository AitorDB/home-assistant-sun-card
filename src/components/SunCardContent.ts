import { TemplateResult, html } from 'lit-html'
import { TSunCardData, TSunCardConfig } from '../types'
import { SunCardFooter } from './SunCardFooter'
import { SunCardGraph } from './SunCardGraph'
import { SunCardHeader }  from './SunCardHeader'
import utils from '../utils'

export class SunCardContent {

    private config: TSunCardConfig
    private data: TSunCardData
  
    constructor (config: TSunCardConfig, data: TSunCardData) {
      this.config = config
      this.data = data
    }

  render (): TemplateResult {
    return html`
      <ha-card>
        <div class="sun-card ${this.config.darkMode ? 'sun-card-dark' : ''}">
          ${ this.showHeader() ? this.renderHeader() : utils.nothing }
          ${ this.renderGraph() }
          ${ this.showFooter() ? this.renderFooter() : utils.nothing }
        </div>
      </ha-card>
    `
  }

  private renderHeader (): TemplateResult {
    return new SunCardHeader(this.config, this.data).render()
  }

  private renderGraph (): TemplateResult {
    return new SunCardGraph(this.data).render()
  }

  private renderFooter (): TemplateResult {
    return new SunCardFooter(this.config, this.data).render()
  }

  private showHeader (): boolean {
      // logic based on config
      return true
  }

  private showFooter (): boolean {
      // logic based on config
      return true
  }
}
