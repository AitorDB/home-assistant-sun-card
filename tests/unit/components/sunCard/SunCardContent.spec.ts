import { html } from 'lit'

import { SunCardContent } from '../../../../src/components/sunCard/SunCardContent'
import { ISunCardConfig, TSunCardData } from '../../../../src/types'
import { HelperFunctions } from '../../../../src/utils/HelperFunctions'
import { TemplateResultTestHelper } from '../../../helpers/TestHelpers'

jest.mock('../../../../src/components/sunCard/SunCardHeader', () => require('../../../mocks/SunCardHeader'))
jest.mock('../../../../src/components/sunCard/SunCardGraph', () => require('../../../mocks/SunCardGraph'))
jest.mock('../../../../src/components/sunCard/SunCardFooter', () => require('../../../mocks/SunCardFooter'))

describe('SunCardContent', () => {
  describe('render', () => {
    beforeAll(() => {
      jest.spyOn(HelperFunctions, 'nothing').mockImplementation(() => html``)
    })

    it('sets dark mode when it is set to true on the config', async () => {
      const config: ISunCardConfig = {
        type: 'sun-card',
        darkMode: true
      }

      const sunCardContent = new SunCardContent(config, {} as TSunCardData)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not set dark mode when it is set to false on the config', async () => {
      const config: ISunCardConfig = {
        type: 'sun-card',
        darkMode: false
      }

      const sunCardContent = new SunCardContent(config, {} as TSunCardData)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the header when show header returns true', async () => {
      const sunCardContent = new SunCardContent({} as ISunCardConfig, {} as TSunCardData)
      jest.spyOn((sunCardContent as unknown as { showHeader: () => boolean }), 'showHeader').mockImplementation(() => true)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not render the header when show header returns false', async () => {
      const sunCardContent = new SunCardContent({} as ISunCardConfig, {} as TSunCardData)
      jest.spyOn((sunCardContent as unknown as { showHeader: () => boolean }), 'showHeader').mockImplementation(() => false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the footer when show footer returns true', async () => {
      const sunCardContent = new SunCardContent({} as ISunCardConfig, {} as TSunCardData)
      jest.spyOn((sunCardContent as unknown as { showFooter: () => boolean }), 'showFooter').mockImplementation(() => true)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not render the footer when show footer returns false', async () => {
      const sunCardContent = new SunCardContent({} as ISunCardConfig, {} as TSunCardData)
      jest.spyOn((sunCardContent as unknown as { showFooter: () => boolean }), 'showFooter').mockImplementation(() => false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof sunCardContent.render>
      element.templateResultFunction = () => sunCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })
})
