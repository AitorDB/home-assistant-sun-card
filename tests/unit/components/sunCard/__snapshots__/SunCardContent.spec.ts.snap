// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SunCardContent render does not render the footer when show footer returns false 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit--><!--?-->
        </div>
      </ha-card>
    
`;

exports[`SunCardContent render does not render the header when show header returns false 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit--><!--?-->
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`SunCardContent render does not set dark mode when it is set to false on the config 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`SunCardContent render renders the footer when show footer returns true 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`SunCardContent render renders the header when show header returns true 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`SunCardContent render sets dark mode when it is set to true on the config 1`] = `
<!---->
      <ha-card>
        <div class="sun-card sun-card-dark">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;
